ctru-rs/src/applets/swkbd.rs

```diff
@@ -675,6 +640,292 @@ impl SoftwareKeyboard {
         // Activate the specific validation rule for maximum length.
         self.state.valid_input = ValidInput::FixedLen.into();
     }
+
+    // A reimplementation of `swkbdInputText` from `libctru/source/applets/swkbd.c`. Allows us to
+    // get text from the software keyboard and put it directly into a `String` without requiring
+    // an intermediate fixed-size buffer
+    fn swkbd_input_text(&mut self, output: &mut String) -> SwkbdButton {
+        use ctru_sys::{
+            MEMPERM_READ, MEMPERM_WRITE, R_FAILED, SWKBD_BUTTON_LEFT, SWKBD_BUTTON_MIDDLE,
+            SWKBD_BUTTON_NONE, SWKBD_BUTTON_RIGHT, SWKBD_D0_CLICK, SWKBD_D1_CLICK0,
+            SWKBD_D1_CLICK1, SWKBD_D2_CLICK0, SWKBD_D2_CLICK1, SWKBD_D2_CLICK2,
+            SWKBD_FILTER_CALLBACK, SWKBD_OUTOFMEM,
+        };
+
+        let swkbd = self.state.as_mut();
+        let mut extra = unsafe { swkbd.__bindgen_anon_1.extra };
+
+        // Calculate shared mem size
+        let mut shared_mem_size = 0;
+
+        shared_mem_size +=
+            (std::mem::size_of::<u16>() * (swkbd.max_text_len as usize + 1) + 3) & !3;
+
+        let dict_off = shared_mem_size;
+
+        shared_mem_size +=
+            (std::mem::size_of::<SwkbdDictWord>() * swkbd.dict_word_count as usize + 3) & !3;
+
+        let status_off = shared_mem_size;
+
+        shared_mem_size += if swkbd.initial_learning_offset >= 0 {
+            std::mem::size_of::<SwkbdStatusData>()
+        } else {
+            0
+        };
+
+        let learning_off = shared_mem_size;
+
+        shared_mem_size += if swkbd.initial_learning_offset >= 0 {
+            std::mem::size_of::<SwkbdLearningData>()
+        } else {
+            0
+        };
+
+        if swkbd.save_state_flags & (1 << 0) != 0 {
+            swkbd.status_offset = shared_mem_size as _;
+            shared_mem_size += std::mem::size_of::<SwkbdStatusData>();
+        }
+
+        if swkbd.save_state_flags & (1 << 1) != 0 {
+            swkbd.learning_offset = shared_mem_size as _;
+            shared_mem_size += std::mem::size_of::<SwkbdLearningData>();
+        }
+
+        shared_mem_size = (shared_mem_size + 0xFFF) & !0xFFF;
+
+        swkbd.shared_memory_size = shared_mem_size;
+
+        // Allocate shared mem
+        let swkbd_shared_mem_ptr = unsafe { libc::memalign(0x1000, shared_mem_size) };
```

Comment:
huh, I was thinking it would be nice to use Rust `alloc` for this but I guess it's passed to C immediately anyway so there's not really a compelling reason to do that that I can think of.

Reply:
I did think about using `std::alloc`, but as you said the memory block is only being manipulated by C code anyway and so it's more straightforward to just use `memalign` and `free`